File: RaftSwift/Sources/RaftCore/Node/Persistence/FileRaftNodePersistence.swift
import Foundation

/// File based implementation of node persistence.
public actor FileRaftNodePersistence: RaftNodePersistence {
    private let directoryURL: URL
    public let compactionThreshold: Int

    public init(compactionThreshold: Int) throws {
        self.compactionThreshold = compactionThreshold
        let fileManager = FileManager.default
        let path = "raft-snapshots"

        // Relative path
        directoryURL = URL(fileURLWithPath: fileManager.currentDirectoryPath)
            .appendingPathComponent(path, isDirectory: true)

        // Ensure the directory exists
        try fileManager.createDirectory(at: directoryURL, withIntermediateDirectories: true)
    }

    private func fileURL(for nodeId: Int) -> URL {
        directoryURL.appendingPathComponent("snapshot_\(nodeId).json")
    }

    public func saveSnapshot(_ snapshot: Snapshot, for nodeId: Int) async throws {
        let url = fileURL(for: nodeId)
        let data = try JSONEncoder().encode(snapshot)

        try data.write(to: url, options: .atomic)

        print("Saved snapshot to \(url.path)")
    }

    public func loadSnapshot(for nodeId: Int) async throws -> Snapshot? {
        let url = fileURL(for: nodeId)
        guard FileManager.default.fileExists(atPath: url.path) else {
            return nil
        }

        let data = try Data(contentsOf: url)

        return try JSONDecoder().decode(Snapshot.self, from: data)
    }

    public func deleteSnapshot(for nodeId: Int) async throws {
        let url = fileURL(for: nodeId)

        if FileManager.default.fileExists(atPath: url.path) {
            try FileManager.default.removeItem(at: url)
        }
    }
}

File: RaftKotlin/raft-node/src/main/kotlin/raft/core/node/persistence/FileRaftNodePersistence.kt
package raft.core.node.persistence

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.json.Json
import raft.core.node.RaftNodePersistence
import raft.core.utils.types.Snapshot
import java.io.File
import java.io.IOException

/**
 * File based implementation of node persistence.
 */
class FileRaftNodePersistence(
    override val compactionThreshold: Int
) : RaftNodePersistence {

    private val directoryPath: String
    private val mutex = Mutex()

    init {
        val currentDir = System.getProperty("user.dir")
        directoryPath = "$currentDir/raft-snapshots"

        // Ensure the directory exists
        val directory = File(directoryPath)
        if (!directory.exists()) {
            if (!directory.mkdirs()) {
                throw IOException("Failed to create directory: $directoryPath")
            }
        }
    }

    private fun getFileForNode(nodeId: Int): File {
        return File(directoryPath, "snapshot_$nodeId.json")
    }

    override suspend fun saveSnapshot(snapshot: Snapshot, nodeId: Int) {
        mutex.withLock {
            val file = getFileForNode(nodeId)
            val jsonString = Json.encodeToString(snapshot)

            // Write atomically by writing to a temp file first, then renaming
            val tempFile = File("${file.absolutePath}.tmp")
            try {
                tempFile.writeText(jsonString)
                if (!tempFile.renameTo(file)) {
                    throw IOException("Failed to rename temp file to ${file.name}")
                }
            } catch (e: Exception) {
                // Clean up temp file if it exists
                if (tempFile.exists()) {
                    tempFile.delete()
                }
                throw e
            }
        }
    }

    override suspend fun loadSnapshot(nodeId: Int): Snapshot? {
        mutex.withLock {
            val file = getFileForNode(nodeId)

            if (!file.exists()) {
                return null
            }

            val jsonString = file.readText()
            return Json.decodeFromString<Snapshot>(jsonString)
        }
    }

    override suspend fun deleteSnapshot(nodeId: Int) {
        mutex.withLock {
            val file = getFileForNode(nodeId)

            if (file.exists()) {
                if (!file.delete()) {
                    throw IOException("Failed to delete snapshot file: ${file.name}")
                }
            }
        }
    }
}
